File: Reqnroll/Chapter14/14.7-ValidationTesterPattern/After/WIMP.Specs/Support/OrderObjectMother.cs
using WIMP.App.Models;

namespace WIMP.Specs.Support;

public class OrderObjectMother
{
    private readonly Order order = new();

    public OrderObjectMother()
    {
        // add default item
        order.AddItem(DomainDefaults.PizzaItemDefaultInstance());
        // set default collection
        order.Collection = DomainDefaults.OrderCollection;
    }

    public Order Build()
    {
        return order;
    }

    public OrderObjectMother WithCustomerCollection()
    {
        order.Collection = OrderCollection.CustomerCollection;
        return this;
    }

    public OrderObjectMother WithAdditionalItem(
        string? name = null, PizzaSize? size = null, PizzaStyle? style = null)
    {
        var pizzaItem = DomainDefaults.PizzaItemDefaultInstance(name, size, style);
        order.AddItem(pizzaItem);
        return this;
    }

    public OrderObjectMother WithItem(
        string? name = null, PizzaSize? size = null, PizzaStyle? style = null)
    {
        order.ClearItems(); // remove existing items
        return WithAdditionalItem(name, size, style);
    }

    public OrderObjectMother WithItems(int quantity,
        string? name = null, PizzaSize? size = null, PizzaStyle? style = null)
    {
        order.ClearItems(); // remove existing items
        for (int i = 0; i < quantity; i++)
        {
            WithAdditionalItem(name, size, style);
        }
        return this;
    }
}

File: CucumberJVM/Chapter14/14.7-ValidationTesterPattern/After/src/test/java/com/wimp/app/specs/support/OrderObjectMother.java
package com.wimp.app.specs.support;

import com.wimp.app.models.*;

public class OrderObjectMother {
    private final Order order;

    public OrderObjectMother() {
        order = new Order();
        // add default item
        order.addItem(DomainDefaults.pizzaItemDefaultInstance());
        order.setDeliveryAddress(DomainDefaults.CUSTOMER_ADDRESS);
    }

    public Order build() {
        return order;
    }

    public OrderObjectMother withAdditionalItem(String name, PizzaSize size, PizzaStyle style) {
        PizzaItem pizzaItem = DomainDefaults.pizzaItemDefaultInstance(name, size, style);
        order.addItem(pizzaItem);
        return this;
    }

    public OrderObjectMother withItem(String name, PizzaSize size, PizzaStyle style) {
        order.clearItems(); // remove existing items
        return withAdditionalItem(name, size, style);
    }

    public OrderObjectMother withItem(String name) {
        return withItem(name, null, null);
    }

    public OrderObjectMother withItems(int quantity, String name, PizzaSize size, PizzaStyle style) {
        order.clearItems(); // remove existing items
        for (int i = 0; i < quantity; i++) {
            withAdditionalItem(name, size, style);
        }
        return this;
    }

    public OrderObjectMother withItems(int quantity, PizzaSize size) {
        return withItems(quantity, null, size, null);
    }

    public OrderObjectMother withCustomerCollection() {
        order.setDeliveryMethod(DeliveryMethod.CUSTOMER_COLLECTION);
        return this;
    }
}
